### ISSUE TITLE
INSTALL: MacOS PETSc error on install

### ISSUE THREAD

**Issue by dthillaithevan** [firedrakeproject/firedrake#3123]
**Describe the error**
Installing firedrake using pyenv (and homebrew) on MacOS 13.6 following [these](https://github.com/firedrakeproject/firedrake/wiki/Installing-pyenv) instructions is giving a PETSc error.  

**Steps to Reproduce**

```
brew install pyenv
export PYENV_ROOT=$HOME/.pyenv
export PATH=$PYENV_ROOT/bin:$PATH
eval "$(pyenv init -)"
pyenv install 3.11
pyenv shell 3.11
curl -O https://raw.githubusercontent.com/firedrakeproject/firedrake/master/scripts/firedrake-install
python firedrake-install
# Below command also fails
# python firedrake-install --package-branch petsc ksagiyam/merge_upstream
```

**Expected behaviour**
Installation without errors

**Error message**
```
Depending on your platform, PETSc may take an hour or more to build!
Traceback (most recent call last):
  File "/Users/dt718/firedrake-install", line 1887, in <module>
    install("petsc/")
  File "/Users/dt718/firedrake-install", line 1043, in install
    build_and_install_petsc()
  File "/Users/dt718/firedrake-install", line 1155, in build_and_install_petsc
    check_call([python, "./configure", "PETSC_DIR={}".format(petsc_dir), "PETSC_ARCH={}".format(petsc_arch)] + petsc_options)
  File "/Users/dt718/firedrake-install", line 666, in check_call
    log.debug(subprocess.check_output(arguments, stderr=subprocess.STDOUT, env=os.environ).decode())
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dt718/.pyenv/versions/3.11.5/lib/python3.11/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dt718/.pyenv/versions/3.11.5/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/Users/dt718/firedrake/bin/python', './configure', 'PETSC_DIR=/Users/dt718/firedrake/src/petsc', 'PETSC_ARCH=default', '--download-pnetcdf', '--download-mpich', '--download-pastix', '--download-ml', '--LDFLAGS=-Wl,-rpath,/opt/homebrew/opt/openblas/lib -L/opt/homebrew/opt/openblas/lib', '--with-zlib', '--with-c2html=0', '--download-hypre', '--download-ptscotch', '--download-scalapack', '--download-hwloc-configure-arguments=--disable-opencl', '--download-chaco', '--download-bison', '--download-hwloc', '--with-blaslapack-dir=/opt/homebrew/opt/openblas', '--download-hdf5', '--CFLAGS=-I/opt/homebrew/opt/openblas/include -Wno-implicit-function-declaration', '--with-x=0', '--download-mpich-configure-arguments=--disable-opencl', '--download-superlu_dist', '--download-mumps', '--with-debugging=0', '--with-shared-libraries=1', '--download-netcdf', '--with-fortran-bindings=0', '--download-metis', '--download-suitesparse']' returned non-zero exit status 1.
```

[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/12725376/firedrake-install.log)
[configure.log](https://github.com/firedrakeproject/firedrake/files/12725371/configure.log)

**Environment:**
 - OS: MacOS 13.6 
 - Python version: 3.11.5
 - pyenv: 2.3.27
 - Homebrew: 4.1.13

**Comment by connorjward** [comment]
Please can you (in order):
1. Update homebrew so that the command `brew doctor` no longer spews warnings.
2. Check that you can compile and run a simple C hello world program.
3. Run `firedrake-install` with the `--package-manager` flag as before but also pass `--with-blas=download`. I have a suspicion that that might help.

**Comment by dthillaithevan** [comment]
Thanks Connor! I ran the following and am getting further in the installation process, it is now failing on `libsupermesh`

``` 
brew update      
gcc -o hello_world_c hello_world_c.c
./hello_world_c # prints Hello world!
curl -O https://raw.githubusercontent.com/firedrakeproject/firedrake/master/scripts/firedrake-install
python firedrake-install --package-branch petsc ksagiyam/merge_upstream --with-blas=download 
```

Error message: 

```
Successfully cloned repository libsupermesh
Checking out branch master
Successfully checked out branch master
Updating submodules.
Successfully updated submodules.
Traceback (most recent call last):
  File "/Users/dt718/firedrake-install", line 1925, in <module>
    build_and_install_libsupermesh()
  File "/Users/dt718/firedrake-install", line 1274, in build_and_install_libsupermesh
    check_call(["make"])
  File "/Users/dt718/firedrake-install", line 666, in check_call
    log.debug(subprocess.check_output(arguments, stderr=subprocess.STDOUT, env=os.environ).decode())
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dt718/.pyenv/versions/3.11.5/lib/python3.11/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dt718/.pyenv/versions/3.11.5/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['make']' returned non-zero exit status 2.
```

[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/12726956/firedrake-install.log)
[configure.log](https://github.com/firedrakeproject/firedrake/files/12726958/configure.log)

**Comment by connorjward** [comment]
It appears that our libsupermesh repository is sorely in need of an update. You are running with a new compiler and hitting places where some things are deprecated and removed from the standard library.

We will discuss this at the Firedrake meeting tomorrow and hopefully fix this ASAP. In the meantime would you like to try using one of our [Docker containers](https://hub.docker.com/r/firedrakeproject/firedrake)?

**Comment by dthillaithevan** [comment]
Ah I see, thanks for looking into this. 

I'm going to be busy for rest of the week and have unfortunately never had any success with Docker. For the time being I am happy to wait for a fix.

Cheers,
D

**Comment by JDBetteridge** [comment]
We updated `libsupermesh` can you try `firedrake-install --package-branch petsc ksagiyam/merge_upstream --with-blas=download` and let us know if that build. The PETSc branch will be updated in due course,, but is waiting on some additional fixes.

**Comment by boutsitron** [comment]
I had the same problem and I managed to install firedrake using the latest firedrake-install, and running essentially this 

`firedrake-install --package-branch petsc ksagiyam/merge_upstream --with-blas=download`

However, this leads to a broken installation as the tests do not pass. Following the instructions [here](https://www.firedrakeproject.org/download.html), I got the following

```
❯ cd $VIRTUAL_ENV/src/firedrake
❯ pytest tests/regression/ -k "poisson_strong or stokes_mini or dg_advection"
0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
ImportError while loading conftest '/Users/boutsitron/firedrake/src/firedrake/tests/conftest.py'.
tests/conftest.py:7: in <module>
    from firedrake.utils import complex_mode
firedrake/__init__.py:72: in <module>
    from firedrake.assemble import *
firedrake/assemble.py:16: in <module>
    from firedrake import (extrusion_utils as eutils, matrix, parameters, solving,
firedrake/solving.py:24: in <module>
    import firedrake.linear_solver as ls
firedrake/linear_solver.py:2: in <module>
    import firedrake.function as function
firedrake/function.py:14: in <module>
    from firedrake import functionspaceimpl
firedrake/functionspaceimpl.py:18: in <module>
    from firedrake.functionspacedata import get_shared_data, create_element
firedrake/functionspacedata.py:32: in <module>
    from firedrake import mesh as mesh_mod
firedrake/mesh.py:29: in <module>
    from firedrake.adjoint import MeshGeometryMixin
firedrake/adjoint/__init__.py:1: in <module>
    from firedrake.adjoint.function import *               # noqa: F401
firedrake/adjoint/function.py:5: in <module>
    from firedrake.adjoint.blocks import FunctionAssignBlock, ProjectBlock, FunctionSplitBlock, FunctionMergeBlock, SupermeshProjectBlock
firedrake/adjoint/blocks.py:1: in <module>
    from dolfin_adjoint_common.compat import compat
E   ModuleNotFoundError: No module named 'dolfin_adjoint_common'
```

**Comment by connorjward** [comment]
Are you using a different branch of Firedrake? Those files have been deleted in `master`, hence the error.

**Comment by boutsitron** [comment]
Yeah, to be exact, since I need a previous Firedrake version, I installed exactly using

`python firedrake-install --doi 10.5281/zenodo.7801478 --slepc --package-branch petsc ksagiyam/merge_upstream --with-blas=download`

**Comment by connorjward** [comment]
Unfortunately the Zenodo part of Firedrake is notoriously fragile. You may be using the right branch of Firedrake but there have been breaking changes with a number of our dependencies recently. You are observing that pyadjoint has recently had a big refactor.

You could try to fix this by checking out an older pyadjoint commit (in `$VIRTUAL_ENV/src/pyadjoint`), but things might not work.

**Comment by dthillaithevan** [comment]
Hi @JDBetteridge, apologies for the slow reply and thanks for this! 

I have managed to install firedrake without errors but I am getting a mpicc error when I try to run a simple test example. 

Is this due to the code being out-of-date or an underlying bug? 

Thanks!

```

  File ~/Downloads/test.py:35
    solve(a == L, uh, bcs=bc, solver_parameters={"ksp_monitor": None})

  File petsc4py/PETSc/Log.pyx:188 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File petsc4py/PETSc/Log.pyx:189 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File ~/firedrake/src/firedrake/firedrake/adjoint_utils/solving.py:57 in wrapper
    output = solve(*args, **kwargs)

  File ~/firedrake/src/firedrake/firedrake/solving.py:128 in solve
    _solve_varproblem(*args, **kwargs)

  File ~/firedrake/src/firedrake/firedrake/solving.py:164 in _solve_varproblem
    solver.solve()

  File petsc4py/PETSc/Log.pyx:188 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File petsc4py/PETSc/Log.pyx:189 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File ~/firedrake/src/firedrake/firedrake/adjoint_utils/variational_solver.py:89 in wrapper
    out = solve(self, **kwargs)

  File ~/firedrake/src/firedrake/firedrake/variational_solver.py:264 in solve
    dbc.apply(self._problem.u)

  File petsc4py/PETSc/Log.pyx:188 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File petsc4py/PETSc/Log.pyx:189 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File ~/firedrake/src/firedrake/firedrake/adjoint_utils/dirichletbc.py:32 in wrapper
    ret = apply(self, *args, **kwargs)

  File ~/firedrake/src/firedrake/firedrake/bcs.py:431 in apply
    r.assign(self.function_arg, subset=self.node_set)

  File ~/firedrake/src/firedrake/firedrake/bcs.py:297 in function_arg
    self._function_arg_update()

  File petsc4py/PETSc/Log.pyx:188 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File petsc4py/PETSc/Log.pyx:189 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File ~/firedrake/src/firedrake/firedrake/adjoint_utils/interpolate.py:22 in wrapper
    output = interpolate(interpolator, *function, **kwargs)

  File ~/firedrake/src/firedrake/firedrake/interpolation.py:648 in interpolate
    assembled_interpolator = self.callable()

  File ~/firedrake/src/firedrake/firedrake/interpolation.py:813 in callable
    l()

  File ~/firedrake/src/PyOP2/pyop2/parloop.py:207 in compute
    self()

  File petsc4py/PETSc/Log.pyx:188 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File petsc4py/PETSc/Log.pyx:189 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File ~/firedrake/src/PyOP2/pyop2/parloop.py:217 in __call__
    self._compute(self.iterset.core_part)

  File ~/firedrake/src/PyOP2/pyop2/parloop.py:198 in _compute
    self.global_kernel(self.comm, part.offset, part.offset+part.size, *self.arglist)

  File ~/firedrake/src/PyOP2/pyop2/global_kernel.py:330 in __call__
    func = self.compile(comm)

  File petsc4py/PETSc/Log.pyx:188 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File petsc4py/PETSc/Log.pyx:189 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File ~/firedrake/src/PyOP2/pyop2/global_kernel.py:401 in compile
    return compilation.load(self, extension, self.name,

  File ~/firedrake/src/PyOP2/pyop2/compilation.py:605 in load
    compiler = sniff_compiler(exe)

  File ~/firedrake/src/PyOP2/pyop2/compilation.py:95 in sniff_compiler
    output = subprocess.run(

  File ~/.pyenv/versions/3.9.18/lib/python3.9/subprocess.py:505 in run
    with Popen(*popenargs, **kwargs) as process:

  File ~/.pyenv/versions/3.9.18/lib/python3.9/subprocess.py:951 in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,

  File ~/.pyenv/versions/3.9.18/lib/python3.9/subprocess.py:1837 in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)

FileNotFoundError: [Errno 2] No such file or directory: 'mpicc'
```

MWE: 

```

from firedrake import *
length = 1
width = 0.2
mesh = RectangleMesh(40, 20, length, width)
V = VectorFunctionSpace(mesh, "Lagrange", 1)
bc = DirichletBC(V, Constant([0, 0]), 1)
rho = Constant(0.01)
g = Constant(1)
f = as_vector([0, -rho*g])
mu = Constant(1)
lambda_ = Constant(0.25)
Id = Identity(mesh.geometric_dimension()) # 2x2 Identity tensor

def epsilon(u):
    return 0.5*(grad(u) + grad(u).T)

def sigma(u):
    return lambda_*div(u)*Id + 2*mu*epsilon(u)

u = TrialFunction(V)
v = TestFunction(V)
a = inner(sigma(u), epsilon(v))*dx
L = dot(f, v)*dx

uh = Function(V)
solve(a == L, uh, bcs=bc, solver_parameters={"ksp_monitor": None})

displaced_coordinates = interpolate(SpatialCoordinate(mesh) + uh, V)
displaced_mesh = Mesh(displaced_coordinates)
```

**Comment by connorjward** [comment]
Can you try setting the environment variable `PYOP2_CC` to the location of your MPI C compiler. E.g.:
```
$ export PYOP2_CC=$VIRTUAL_ENV/src/petsc/default/bin/mpicc
```
is the usual place.

**Comment by dthillaithevan** [comment]
Thanks Conner - that has fixed the `mpicc` error! I tried the interpolate function which is giving a different error - do you know what could be causing this issue?

```
X = SpatialCoordinate(mesh)
XU = interpolate(X, V)
```

```

  File ~/firedrake/src/firedrake/firedrake/interpolation.py:645 in interpolate
    assembled_interpolator = self.frozen_assembled_interpolator

AttributeError: 'SameMeshInterpolator' object has no attribute 'frozen_assembled_interpolator'


During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File ~/firedrake/src/PyOP2/pyop2/global_kernel.py:328 in __call__
    func = self._func_cache[key]

KeyError: 4944589424


During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File ~/firedrake/src/PyOP2/pyop2/compilation.py:335 in get_so
    return ctypes.CDLL(soname)

  File ~/.pyenv/versions/3.9.18/lib/python3.9/ctypes/__init__.py:374 in __init__
    self._handle = _dlopen(self._name, mode)

OSError: dlopen(/Users/dt718/firedrake/.cache/pyop2/de/53826e7411d1b86d97f2185d255599.so, 0x0006): tried: '/Users/dt718/firedrake/.cache/pyop2/de/53826e7411d1b86d97f2185d255599.so' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/dt718/firedrake/.cache/pyop2/de/53826e7411d1b86d97f2185d255599.so' (no such file), '/Users/dt718/firedrake/.cache/pyop2/de/53826e7411d1b86d97f2185d255599.so' (no such file)


During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File ~/firedrake/lib/python3.9/site-packages/spyder_kernels/py3compat.py:356 in compat_exec
    exec(code, globals, locals)

  File ~/Downloads/test.py:40
    XU = interpolate(X, V)

  File petsc4py/PETSc/Log.pyx:188 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File petsc4py/PETSc/Log.pyx:189 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File ~/firedrake/src/firedrake/firedrake/interpolation.py:138 in interpolate
    return Interpolator(

  File petsc4py/PETSc/Log.pyx:188 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File petsc4py/PETSc/Log.pyx:189 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File ~/firedrake/src/firedrake/firedrake/adjoint_utils/interpolate.py:22 in wrapper
    output = interpolate(interpolator, *function, **kwargs)

  File ~/firedrake/src/firedrake/firedrake/interpolation.py:648 in interpolate
    assembled_interpolator = self.callable()

  File ~/firedrake/src/firedrake/firedrake/interpolation.py:813 in callable
    l()

  File ~/firedrake/src/PyOP2/pyop2/parloop.py:207 in compute
    self()

  File petsc4py/PETSc/Log.pyx:188 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File petsc4py/PETSc/Log.pyx:189 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File ~/firedrake/src/PyOP2/pyop2/parloop.py:217 in __call__
    self._compute(self.iterset.core_part)

  File ~/firedrake/src/PyOP2/pyop2/parloop.py:198 in _compute
    self.global_kernel(self.comm, part.offset, part.offset+part.size, *self.arglist)

  File ~/firedrake/src/PyOP2/pyop2/global_kernel.py:330 in __call__
    func = self.compile(comm)

  File petsc4py/PETSc/Log.pyx:188 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File petsc4py/PETSc/Log.pyx:189 in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func

  File ~/firedrake/src/PyOP2/pyop2/global_kernel.py:401 in compile
    return compilation.load(self, extension, self.name,

  File ~/firedrake/src/PyOP2/pyop2/compilation.py:606 in load
    dll = compiler(cppargs, ldargs, cpp=cpp, comm=comm).get_so(code, extension)

  File ~/firedrake/src/PyOP2/pyop2/compilation.py:415 in get_so
    return ctypes.CDLL(soname)

  File ~/.pyenv/versions/3.9.18/lib/python3.9/ctypes/__init__.py:374 in __init__
    self._handle = _dlopen(self._name, mode)

OSError: dlopen(/Users/dt718/firedrake/.cache/pyop2/de/53826e7411d1b86d97f2185d255599.so, 0x0006): tried: '/Users/dt718/firedrake/.cache/pyop2/de/53826e7411d1b86d97f2185d255599.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64')), '/System/Volumes/Preboot/Cryptexes/OS/Users/dt718/firedrake/.cache/pyop2/de/53826e7411d1b86d97f2185d255599.so' (no such file), '/Users/dt718/firedrake/.cache/pyop2/de/53826e7411d1b86d97f2185d255599.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64'))
```

**Comment by dthillaithevan** [comment]
It seems the issue was Spyder - if I run the code in terminal it runs without error

**Comment by Ig-dolci** [comment]
Yep. I had no error here on my Mac as well.

**Comment by connorjward** [comment]
> It seems the issue was Spyder - if I run the code in terminal it runs without error

Can this issue be closed?

**Comment by dthillaithevan** [comment]
Hi @connorjward, yes - I'll close this now.

**Comment by ReHoss** [comment]
I also have the error using PyCharm

**Comment by dham** [comment]
> I also have the error using PyCharm

This issue doesn't have anything to do with PyCharm. If you have an issue using Firedrake please [open a new discussion](https://github.com/firedrakeproject/firedrake/discussions/new/choose) detailing exactly what you did and providing the whole error message.

**Comment by ReHoss** [comment]
> > I also have the error using PyCharm
> 
> This issue doesn't have anything to do with PyCharm. If you have an issue using Firedrake please [open a new discussion](https://github.com/firedrakeproject/firedrake/discussions/new/choose) detailing exactly what you did and providing the whole error message.

This issue helped me to understand the problem did not came from Firedrake but our Integrated Development Environments (IDE).
Thus, this issue informs beginners `FileNotFoundError: [Errno 2] No such file or directory: 'mpicc'` may be linked to the use of IDE. Indeed, in my case the `PATH` used by my run configuration was incomplete and did not incorporate firedrake binaries.
Consequently, there is no need to open a new issue. :)  

Comment from @dthillaithevan that helped me:
> It seems the issue was Spyder - if I run the code in terminal it runs without error

Best,
